### PR TITLE
support composite primary key insert/update/delete

### PR DESCRIPTION
**Pull request by bbbearxyz**
1. support composite primary key insert/update/delete
2. fix primary key allow null value bug

## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #4513 #5162 

## What this PR does / why we need it:

### CODE REVIEW

**Review by daviszhen — APPROVED** [review]

**Review by aunjgr — APPROVED** [review]

File: proto/plan.proto [hunk 1]
```diff
@@ -281,6 +281,7 @@ message TableDef {
 	string table_type 		= 4;
 	string createsql 		= 5;
 	map<string, int32> name2col_index = 6;
+	ColDef composite_pkey = 7;
```
Comment (ouyuanning) [inline proto/plan.proto:284]: pls add some code to DeepCopyTableDef to copy this new field.

File: pkg/sql/plan/build_ddl.go [hunk 2]
```diff
@@ -385,13 +385,37 @@ func buildTableDefs(defs tree.TableDefs, ctx CompilerContext, tableDef *TableDef
...
+					Id:   int32(types.T_varchar),
+					Size: types.VarlenaSize,
+				},
+				Default: &plan.Default{
+					NullAbility:  false,
+					Expr:         nil,
+					OriginString: "",
+				},
+			})
+			tableDef.Defs = append(tableDef.Defs, &plan.TableDef_DefType{
```
Comment (fengttt) [inline pkg/sql/plan/build_ddl.go:411]: This won't work in general, what if user alter table rename a column?    Unless you take care of that fixing table defs.
Do we?

**Review by fengttt — COMMENTED** [review]
Either support checking primary key of alter table rename/drop/nullable, or, file a NotYetImpl bug.

BTW, single column primary key is stored in the same way as composite key, right?

File: pkg/sql/plan/build_ddl.go (hunk 2 above)
Comment (bbbearxyz) [reply pkg/sql/plan/build_ddl.go:411]: I think alter is not implement yet, when alter is implemented, I need to add some code

**Comment by bbbearxyz** [comment]
> Either support checking primary key of alter table rename/drop/nullable, or, file a NotYetImpl bug.
> 
> BTW, single column primary key is stored in the same way as composite key, right?

no, signle primary key is stored in old way.

**Comment by fengttt** [comment]
> > Either support checking primary key of alter table rename/drop/nullable, or, file a NotYetImpl bug.
> > BTW, single column primary key is stored in the same way as composite key, right?
> 
> no, signle primary key is stored in old way.

Why?   Don't you want a consistant way of handling these?

**Comment by bbbearxyz** [comment]
> > > Either support checking primary key of alter table rename/drop/nullable, or, file a NotYetImpl bug.
> > > BTW, single column primary key is stored in the same way as composite key, right?
> > 
> > 
> > no, signle primary key is stored in old way.
> 
> Why? Don't you want a consistant way of handling these?

because old way we don't need to serial, it is faster. If it is single, we don't need to add extra column, tae engine can make true single primary key is correct.

**Review by fengttt — APPROVED** [review]

**Review by aressu1985 — APPROVED** [review]

**Review by nnsgmsone — APPROVED** [review]

**Comment by bbbearxyz** [comment]
@matrix-meow rerun failed

**Pull request merged**